easybuild/easyblocks/q/quantumespresso.py

```diff
@@ -62,6 +63,59 @@ def __init__(self, *args, **kwargs):
 
         self.install_subdir = "espresso-%s" % self.version
 
+    def obtain_file(self, *args, **kwargs):
+        """Custom implementation of obtain_file, to deal with problematic downloads from qe-forge.org."""
+        filename = args[0]
+        source_urls = self.cfg['source_urls']
+
+        # check whether file is already there, by running obtain_file without providing any source_urls
+        path = None
+        self.cfg['source_urls'] = []
+        try:
+            path = super(EB_QuantumESPRESSO, self).obtain_file(*args, **kwargs)
+        except EasyBuildError, err:
+            self.log.debug("Ignoring expected download error: %s", err)
+
+        # restore source_urls value
+        self.cfg['source_urls'] = source_urls
+
+        if path:
+            self.log.info("Source file %s already there at %s, assuming it's OK.", filename, path)
+
+        elif filename.endswith('tar.gz') or filename.endswith('.tgz'):
+            # put copy of provided source_urls in place to play around with
+            kwargs.update({'urls': source_urls[:]})
+
+            # if source is not available yet, try to download it
+            path = super(EB_QuantumESPRESSO, self).obtain_file(*args, **kwargs)
+
+            # check whether downloaded tarball can be unpacked
+            _, ec = run_cmd("tar tfz %s" % path, simple=False, log_ok=False, log_all=False)
+
+            # download may result in an HTML page when wrong source URL was used (rather than a proper 404 HTML code)
```

Comment:
If there's an MD5sum available, it will detect this issue, yes, but simply complain that things don't match.

Checking the MD5sum if you know that the download is faulty is pretty useless.